nouveau/src/main/java/org/apache/couchdb/nouveau/core/IndexManager.java

```diff
@@ -82,43 +114,104 @@ public interface IndexFunction<V, R> {
 
     private ObjectMapper objectMapper;
 
-    private MetricRegistry metricRegistry;
-
-    private Scheduler scheduler;
+    private ScheduledExecutorService schedulerExecutorService;
 
     private SearcherFactory searcherFactory;
 
-    private AsyncLoadingCache<String, Index> cache;
+    private LRUMap cache;
 
-    private StripedLock<String> lock;
+    private StripedLock<String> createLock;
 
     public <R> R with(final String name, final IndexFunction<Index, R> indexFun)
             throws IOException, InterruptedException {
+        retry:
         while (true) {
             if (!exists(name)) {
                 throw new WebApplicationException("Index does not exist", Status.NOT_FOUND);
             }
-
-            final CompletableFuture<Index> future = cache.get(name);
-            final Index index = future.join();
-
-            if (index.tryAcquire()) {
-                try {
-                    return indexFun.apply(index);
-                } finally {
-                    index.release();
+            final IndexHolder holder;
+            synchronized (cache) {
+                holder = cache.computeIfAbsent(name, (k) -> new IndexHolder());
+            }
+            holder.lock.readLock().lock();
+            try {
+                // Load if not already loaded.
+                if (holder.state == HolderState.NOT_LOADED) {
+                    holder.lock.readLock().unlock();
+                    holder.lock.writeLock().lock();
+                    try {
+                        if (holder.state == HolderState.NOT_LOADED) {
+                            holder.index = load(name);
+                            holder.commitFuture = this.schedulerExecutorService.scheduleWithFixedDelay(
+                                    commitFun(name, holder),
+                                    commitIntervalSeconds,
+                                    commitIntervalSeconds,
+                                    TimeUnit.SECONDS);
+                            holder.state = HolderState.LOADED;
+                        }
+                    } finally {
+                        holder.lock.readLock().lock();
```

Comment:
If we hold the write lock we're allowed to take a read lock and now we hold both?

I wonder if ` holder.lock.readLock().lock();` should be before the `} finally {` instead of after, in other words `finally` should only handle unlocking the write lock we took above in line 141 ` holder.lock.writeLock()`?

Reply:
Yes, ```
Lock downgrading

Reentrancy also allows downgrading from the write lock to a read lock, by acquiring the write lock, then the read lock and then releasing the write lock. However, upgrading from a read lock to the write lock is not possible. 
```
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/ReentrantReadWriteLock.html

Reply:
the purpose here is to ensure we have a read lock or write lock at all times.

Reply:
I was thinking if `readLock().lock()` could fail (does it throw or get interrupted?) then the next statement `holder.lock.writeLock().unlock();` won't be run and it would be stuck with a write lock taken.

Reply:
ah, I see what you mean. will ponder. the proper spell for this in is those javadocs, will double check

Reply:
updated. agree that the read lock() should be before the finally for the reason stated. pattern now matches the javadoc example more closely.